k2/csrc/ragged_ops.cu

```diff
@@ -1562,4 +1563,125 @@ RaggedShape GetLayer(const RaggedShape &src, int32_t layer) {
 }
 
 
+void DecomposeRaggedShape(const RaggedShape &src,
+                          int32_t axis,
+                          RaggedShape *top, RaggedShape *bottom) {
+  K2_CHECK_GT(axis, 0);
+  K2_CHECK_LT(axis, src.NumAxes() - 1);
+  const std::vector<RaggedShapeLayer> &src_layers = src.Layers();
+  std::vector<RaggedShapeLayer> top_layers(axis),
+      bottom_layers(src_layers.size() - axis);
+  int32_t src_size = src_layers.size();
+  for (int32_t i = 0; i < axis; ++i)
+    top_layers[i] = src_layers[i];
+  for (int32_t i = axis; i < src_size; ++i)
+    bottom_layers[i - axis] = src_layers[i];
+  *top = RaggedShape(top_layers);
+  *bottom = RaggedShape(bottom_layers);
+}
+
+RaggedShape RemoveEmptyLists(RaggedShape &src_shape,
+                             int32_t axis,
+                             Renumbering *renumbering_out) {
+  if (axis == 0) {
+    return RemoveEmptyListsAxis0(src_shape, renumbering_out);
+  }
+  RaggedShape top_shape, bottom_shape;
+  DecomposeRaggedShape(src_shape, axis, &top_shape, &bottom_shape);
+
+  Renumbering r_temp;
+  if (!renumbering_out)
+    renumbering_out = &r_temp;
+  bottom_shape = RemoveEmptyListsAxis0(bottom_shape, renumbering_out);
+  top_shape = SubsampleRaggedShape(top_shape, *renumbering_out);
+  return ComposeRaggedShapes(top_shape, bottom_shape);
+}
+
+
+RaggedShape RemoveSomeEmptyLists(RaggedShape &src_shape,
+                                 int32_t axis,
+                                 Renumbering &renumbering) {
+  if (axis == 0) {
+    return RenumberAxis0Simple(src_shape, renumbering);
+  }
+  RaggedShape top_shape, bottom_shape;
+  DecomposeRaggedShape(src_shape, axis, &top_shape, &bottom_shape);
+
+  bottom_shape = RenumberAxis0Simple(bottom_shape, renumbering);
+  top_shape = SubsampleRaggedShape(top_shape, renumbering);
+  return ComposeRaggedShapes(top_shape, bottom_shape);
+}
+
+
+
+RaggedShape RemoveEmptyListsAxis0(RaggedShape &src_shape,
+                                  Renumbering *renumbering_out) {
+  Renumbering r_temp;
+  if (!renumbering_out)
+    renumbering_out = &r_temp;
+
+  ContextPtr c = src_shape.Context();
+  int32_t num_lists = src_shape.Dim0();
+  *renumbering_out = Renumbering(c, num_lists);
+  int32_t *row_splits_data = src_shape.RowSplits(1).Data();
+  char *keep_data = renumbering_out->Keep().Data();
+  K2_EVAL(c, num_lists + 1, lambda_set_keep, (int32_t i) -> void {
```

Comment:
Should this be `num_lists`? 

Otherwise, when `i` is `num_lists`, `row_splits_data[i+1]` is accessed out of bound.

Reply:
I think you are right.   Would you mind making a PR to fix it?